### DISCUSSION TITLE
Thursday Scrum Update

### DISCUSSION THREAD

**Discussion by anlhu** [casa-business-manager/CASA#56] [General]
Since tuesday I have completed the roles integration, but it's waiting on @IanChong1 to review before merging, so it's not done yet.

There seems to be an issue in the unit test clean up where I can't delete users because of foreign keys that rely on them (it's probably the table that stores relations from roles to users), but it also won't let me delete the role or the organization because of "Expected [201] but was [400]" which is weird because I'm not even using an API call, I'm just calling `OrganizationRepository.deleteById()`. For now I just changed users to cascade delete their roles so the unit tests pass. I'm pretty sure this is just something weird in the unit testing so before we go to prod we can just revert that.

Before next time I'm not sure what to do.
I don't really know how to set up permissions or enforcement for the roles. I would guess somehow its middleware but I don't know where that is stored. I also can't access the scrum board or see what you guys have been up to.
@catdiehl could you post those here for me? Then we can talk about what task I can pick up next.

My blocker for the last task was the unit testing thing, but I have just patched it with a temp fix for now.

**Comment by catdiehl** [comment]
I pushed my work for today into roles-permissions, changed permissions from a string to a set of permission objects and the necessary implementation in the backend. I am hoping for no tricky back-reference issues but you never know with Java backend :).

I need a way to implement them in the frontend, I am going to make an API function that returns the current user's permissions as a set so that we can easily compare the necessary permissions, especially in actions that may require multiple perms such as sending a message to a user higher in the hierarchy, open to other ideas but it seems the most efficient for API calls

I updated RoleTest from permissions being a string but we will need permission test since all the tests currently just set the role to all perms T_T

**Reply by catdiehl** [reply]
pushed frontend function with connector in RoleController.java

I broke the hell out of unit tests they all return 403 on expected 200 I dont have the energy to do it right now I just wanted to put everything I needed to out on code in the branch and will revise later

**Reply by anlhu** [reply]
Unit tests will have to be adjusted for the new DTO since permissions are a set of strings now but that con be done when the backend is done. 

Also wouldnt a dictionary be better to map from permission to Boolean? If not maybe you can define permissions in an enum so we are mapping from numbers to booleans and that can just be in an array. Something like 
```java 
enum PERMISSIONS {
MANAGE_ROLES: 0,
VIEW_ROLES: 1,
MANAGE_ORG: 2,
etc...
}

Role.permissions = new Boolean[Size];

Role.permissions[PERMISSIONS.MANAGE_ROLES] = true;
```

**Reply by catdiehl** [reply]
just saw this now, java maps the set to a flatMap which it can use to check if the necessary perms are present for an action. the perms are defined in Permission.java
